src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -162,6 +508,105 @@ CQueryMutators::NormalizeGroupByProjList(CMemoryPool *mp,
 {
 	Query *query_copy = (Query *) gpdb::CopyObject(const_cast<Query *>(query));
 
+	// If there is a functional dependency proved for a relation on a set of
+	// grouping columns, then it is valid to have a column in the target list
+	// even if it is not listed explicitly in the groupClause. To make ORCA
+	// correctly process such cases, we add all such target list entries into
+	// groupClause's grouping sets, where the functional dependency exists.
+	// If there is such functional dependency existing, the OID of
+	// Primary Key constraint resides in the constraintDeps list. As the
+	// constraintDeps list contains only Primary Key OIDs, we skip checks of
+	// constraint type in all code below.
+	// Refer to function check_functional_grouping for more details.
+	if (0 != gpdb::ListLength(query_copy->constraintDeps))
+	{
+		// Explicit addition of functionally dependent columns requires the
+		// following steps:
+		// 1. Store current unique TargetlistEntry references in groupClause -
+		// they will be needed at step 3.
+		// 2. Update all grouping sets that contain Primary Key
+		// 3. Update arguments of GROUPING functions, because they could be
+		// shifted after step 3.
+
+		// Step 1.
+		// Store current unique TargetlistEntry references in groupClause.
+		// After modifying groupClause, they will be required to update
+		// arguments inside grouping functions.
+		SContexGetGroupUniqueTleWalker ctx_old_grouping_tle_refs;
+		GetGroupUniqueTargetlistEntriesWalker((Node *) query_copy->groupClause,
+											  &ctx_old_grouping_tle_refs);
+
+		// Step 2.
+		// Update groupClause - add all functionally dependent entries explicitly.
+		ListCell *lc_constraint = NULL;
+		ForEach(lc_constraint, query_copy->constraintDeps)
```

Comment:
And, what happens when we add the SubLink into the TL? Should we handle that case?

```
SELECT b, (select c from (select c) q1) FROM test_table1 GROUP BY a ORDER BY a;
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Query-to-DXL Translation: No attribute entry found due to incorrect normalization of query
```

Reply:
fixed this case